recipes/license_server.rb

```diff
@@ -25,15 +26,20 @@
 data_bag_item = bag.to_hash
 total_available_license_quota = 0
 
-license_groups = data_bag_item.inject('enterprise' => {}) do |hash, (key, value)|
+license_groups = data_bag_item.inject({}) do |hash, (key, value)|
   unless %w[id chef_type data_bag].include? key
     doc = Nokogiri::XML value
-    type = doc.at_xpath('/license/payload/type/text()').to_s
+    sourcetypes = doc.search('//sourcetype').map(&:text).join
+    node.run_state['type'] = doc.at_xpath('/license/payload/type/text()').to_s
+    node.run_state['type'] = "#{node.run_state['type']}_#{Digest::SHA256.hexdigest(sourcetypes).upcase}" if node.run_state['type'] == 'fixed-sourcetype'
     quota = doc.at_xpath('/license/payload/quota/text()').to_s.to_i
     expiration_time = doc.at_xpath('/license/payload/expiration_time/text()').to_s.to_i
-    total_available_license_quota += quota if type == 'enterprise' && expiration_time > Time.now.to_i
-    hash[type] ||= {}
-    hash[type][key] = value
+    total_available_license_quota += quota if (node.run_state['type'] == 'enterprise' || node.run_state['type'] == "fixed-sourcetype_#{Digest::SHA256.hexdigest(sourcetypes).upcase}") && expiration_time > Time.now.to_i
+    unless hash[node.run_state['type']].nil?
+      fail 'Multiple license types are not currently supported' if hash[node.run_state['type']] != value
```

Comment:
Not sure I follow why we are comparing type to license value, which is the actual license. I think we should be comparing each license type with the license types already in the hash (unless the hash is empty) and fail if we find a different license type.

Reply:
Also, instead of updating node.run_state every time, how about we just read the value of type into a variable and use it inside the loop. After the loop, we can assign this variable to node.run_state['type'].

Reply:
Should that looking something like this?
`    unless hash[type].nil?
      fail 'Multiple license types are not currently supported' if hash[type] != type
    end`

Reply:
I converged the node successfully and verified that the server.conf and licenses were created as expected.